Custom health checks: manager, check types, and on-check-failure handling

benhoyt opened the pull request:
This implements custom health checks and the `on-check-failure` logic per the [spec](https://docs.google.com/document/d/1d6-h3UAt2VPUSvlkVF30l8iuDW8raRNkHbp5M6NUo1A/edit).

Basically, you specify the list of custom health checks in a new top-level configuration object named `checks` (see the `README.md` documentation). They can be HTTP, TCP, or exec checks. Then in your service configuration you specify any action you want to perform `on-check-failure` for the various checks.

Here's an example configuration with a single HTTP-based check, and a single service that will be restarted if the check fails (twice in a row):

```yaml
services:
    web-server:
        command: python3 webapp.py --port=8080
        on-check-failure:
            server-up: restart

checks:
    server-up:
        period: 5s
        failures: 2
        http:
            url: http://localhost:8080/is-up
```

It's a good amount of code, though at least half of it is tests. The CLI and APIs are coming in https://github.com/canonical/pebble/pull/86.

I'm pretty happy with how it turned out, particularly the checkers and check manager. However, the wiring up of the check manager and the service manager is a little odd, because they are mutually dependent on one another: we have to tell the check manager about plan updates (from the service manager), and we have to tell the service manager about check failures (from the check manager). Open to suggestions on that or anything else.

hpidcock commented on README.md:
```diff
@@ -239,6 +244,91 @@ services:
+        # the period, and must not be zero. Default is "3s".
+        timeout: <duration>
+
+        # (Optional) Number of times the check must in a row to be considered
```
consider revising this comment

benhoyt replied:
Done, thanks.

hpidcock commented on README.md:
```diff
@@ -239,6 +244,91 @@ services:
+        # failing and to trigger the on-check-failure action. Default 3.
+        failures: <failure threshold>
+
+        # Type-specific check configuration is below. Note that one (and only
```
Consider dropping this comment for my suggestion below.

benhoyt replied:
Done.

hpidcock commented on README.md:
```diff
@@ -239,6 +244,91 @@ services:
+        # one) of the "http", "tcp", or "exec" objects must be specified.
+
+        # Configures an HTTP check, which is successful if a GET to the
+        # specified URL returns a 20x status code.
```
`# Only "http", "tcp", or "exec" can be specified.`

hpidcock commented on README.md:
```diff
@@ -239,6 +244,91 @@ services:
+
+        # Configures a TCP port check, which is successful if the specified
+        # TCP port is listening and we can successfully open it. Nothing is
+        # sent to the port.
```
`# Only "http", "tcp", or "exec" can be specified.`

hpidcock commented on README.md:
```diff
@@ -239,6 +244,91 @@ services:
+            host: <host name>
+
+        # Configures a command execution check, which is successful if running
+        # the specified command returns a zero exit code.
```
`# Only "http", "tcp", or "exec" can be specified.`

hpidcock commented on internal/overlord/checkstate/checkers.go:
```diff
@@ -0,0 +1,147 @@
+	"github.com/canonical/pebble/internal/strutil/shlex"
+)
+
+// httpChecker is a checker that ensures an HTTP GET at a specified URL returns 20x.
```
`an HTTP GET` => `a HTTP GET`

hpidcock commented on internal/overlord/checkstate/checkers.go:
```diff
@@ -0,0 +1,147 @@
+	headers map[string]string
+}
+
+func (c *httpChecker) check(ctx context.Context) error {
```
do these actually belong in a package called "checkstate"

benhoyt replied:
I see what you mean. They could go in their own internal package and the checkers could be exported. However, I think the pragmatic placement in this package is okay -- they're very much related, and it's likely the only place they'll ever be used. If that changes, we can always move them to a separate package.

niemeyer commented on internal/overlord/overlord.go:
```diff
@@ -112,6 +114,14 @@ func New(pebbleDir string, restartHandler restart.Handler, serviceOutput io.Writ
+	o.serviceMgr.AddPlanHandler(o.checkMgr.Configure)
+
+	// Tell service manager about check failures.
+	o.checkMgr.AddFailureHandler(o.serviceMgr.CheckFailure)
```
The terminology here is a bit problematic as "handlers" for us are a very specific and important idea across the entire code base: task handlers. When possible, we should try not to blur these lines to keep our conversations and minds a bit more organized.

How about:
```
o.serviceMgr.NotifyPlanChanged(o.checkMgr.PlanChanged)
o.checkMgr.NotifyCheckFailed(o.serviceMgr.CheckFailed)
```

We can try to more elegantly avoid the entanglement later on, but this looks trivial enough that we shouldn't try to over-engineer a solution just now, as there's more important feature-specific complexity in the PR for us to pay attention to.

benhoyt replied:
Good call, updated.

niemeyer commented on internal/overlord/servstate/manager.go:
```diff
@@ -425,3 +445,19 @@ func (m *ServiceManager) SendSignal(services []string, signal string) error {
 	return nil
 }
+
+// CheckFailure response to a health check failure. If the given check name is
```
"CheckFailure" reads too much like "check this failure" (verb) instead of "a check has failed" (noun). Will be solved by the earlier point, though, assuming there's some agreement around it.

niemeyer commented on internal/overlord/servstate/manager.go:
```diff
@@ -425,3 +445,19 @@ func (m *ServiceManager) SendSignal(services []string, signal string) error {
+	for _, service := range m.services {
+		for checkName, action := range service.config.OnCheckFailure {
+			if checkName == name {
+				service.checkFailure(action)
```
This one may need updating as well, though.

niemeyer commented on internal/plan/plan.go:
```diff
@@ -26,18 +26,33 @@ import (
 	"time"
 
+	"github.com/imdario/mergo"
+	"github.com/mitchellh/copystructure"
```
In our conversation I asked whether we could just use reflection, and I thought the positive reply meant the internal reflect package. I didn't realize that by "reflection" you were actually referring to embedding two other packages which introduce plenty of complexity of their own.  Can we please drop these dependencies for this PR, and split out that conversation out?

benhoyt replied:
Ah, I hadn't realized that. I did refer to the two reflect-based libraries in the meeting, but I understand the confusion. I've reverted to the hand-rolled logic for now and we can discuss further in a follow-up PR later.

niemeyer reviewed (changes requested):
Thanks for the PR, Ben. Brilliant work overall.

niemeyer commented:
One more detail for us to discuss before this is final: while reviewing #89, I realized that throughout the code base we already have a term for "halt" well established as "shutdown". I'm not entirely sure this concept has leaked into the UX so far, but I wouldn't be surprised if it already is visible in some part of the API.

In either case, we should probably unify them, in either direction. "shutdown" has the advantage of already being present in the code (so no further patching) and that it's been used by snapd, so we're more likely to get terminology that works across the projects now and in the future.

benhoyt commented on README.md:
```diff
@@ -148,9 +148,7 @@ description: |
 services:
-
     <service name>:
-
```
Fair - fixed.

benhoyt commented on README.md:
```diff
@@ -239,6 +244,98 @@ services:
+# (Optional) A list of health checks managed by this configuration layer.
+checks:
+    <check name>:
+        # (Required) Control how this check definition is combined with any
```
Fixed.

benhoyt commented on README.md:
```diff
@@ -239,6 +244,98 @@ services:
+        # (Optional) Check level, which can be used for filtering checks when
+        # calling the checks API or health endpoint.
+
+        # For the health endpoint, ready implies alive. In other words, if all
```
I've added a `#` comment to the blank line (plenty of other field descriptions have multiple paragraphs, so I think that's okay for a longer description / separation of concepts.

benhoyt commented on README.md:
```diff
@@ -239,6 +244,98 @@ services:
+        # (Optional) Number of times in a row the check must error to be
+        # considered a failure (which triggers the on-check-failure action).
+        # Default 3.
+        failures: <failure threshold>
```
Thanks -- agreed. Fixed.

benhoyt commented on README.md:
```diff
@@ -239,6 +244,98 @@ services:
+            # (Required) Port number to open.
+            port: <port number>
+
+            # (Optional) Hostname to use. Default is "localhost".
```
Yes. Updated.

benhoyt commented on README.md:
```diff
@@ -239,6 +244,98 @@ services:
+            port: <port number>
+
+            # (Optional) Hostname to use. Default is "localhost".
+            host: <host name>
```
Good call. Used "host name" in both places.

benhoyt commented on README.md:
```diff
@@ -239,6 +244,98 @@ services:
+        # Only one of "http", "tcp", or "exec" may be specified.
+        exec:
+            # (Required) Command line to execute. The command is executed
+            # directly; use "/bin/sh -c '...'" to run via the shell.
```
Updated (in both places).

benhoyt commented on go.mod:
```diff
@@ -5,9 +5,11 @@ go 1.14
 require (
 	github.com/gorilla/mux v1.8.0
 	github.com/gorilla/websocket v1.4.2
+	github.com/imdario/mergo v0.3.12
```
No, this is the reflect-based merge library (as you discovered below). Reverted for now and will discuss later in a separate PR.

benhoyt commented on go.sum:
```diff
@@ -11,6 +13,10 @@ github.com/kr/pty v1.1.1/go.mod h1:pFQYn66WHrOpPYNljwOMqo10TkYh1fy3cYio2l3bCsQ=
+github.com/mitchellh/copystructure v1.2.0 h1:vpKXTN4ewci03Vljg/q9QvCGUDttBOGBIa15WveJJGw=
+github.com/mitchellh/copystructure v1.2.0/go.mod h1:qLl+cE2AmVv+CoeAwDPye/v+N2HKCj9FbZEVFJRxO9s=
+github.com/mitchellh/reflectwalk v1.0.2 h1:G2LzWKi524PWgd3mLHV8Y5k7s6XUvT0Gef6zxSIeXaQ=
+github.com/mitchellh/reflectwalk v1.0.2/go.mod h1:mSTlrgnPZtwu0c4WaC2kGObEpuNDbx0jmZXqmk4esnw=
```
No, this is the reflect-based deep copy library (as you discovered below). Reverted for now and will discuss later in a separate PR.

benhoyt commented on internal/overlord/checkstate/checkers.go:
```diff
@@ -0,0 +1,164 @@
+
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		if ctx.Err() == context.DeadlineExceeded {
```
Yeah, Go's `cmd.Start()` starts a goroutine which waits for the context and kills the process if the context is canceled. So this will happen if the check is stopped (due to a config update) or the timeout occurs, and it will be killed / cleaned up properly.

benhoyt commented on internal/overlord/checkstate/checkers.go:
```diff
@@ -0,0 +1,164 @@
+		} else {
+			details = string(output)
+		}
+		return &detailsError{error: err, details: details}
```
OutputErr is a bit weird as it drops the err if there is output. In any case, it wasn't quite what we wanted for the previous service management case (we needed different indentation handling there and a "(...)" prefix if it was truncated). But here it's the same as the servstate case, so I've now made both use a ring buffer and moved that logic to a `servicelog.LastLines` function for reuse.

benhoyt commented on internal/overlord/checkstate/manager.go:
```diff
@@ -0,0 +1,249 @@
+	}
+	c.lastErr = nil
+	c.failures = 0
+	c.actionRan = false
```
I slightly prefer the `if err != nil { ... }` polarity for consistency with basically all other error checks (check error first and return, happy path after). But it is simpler your way, so I've updated it.

benhoyt commented on internal/overlord/servstate/manager.go:
```diff
@@ -64,12 +68,26 @@ func NewManager(s *state.State, runner *state.TaskRunner, pebbleDir string, serv
 
+// AddPlanHandler adds f to the list of "plan handlers", functions that are
+// called whenever the plan is updated. Each function is called with a copy
+// of the plan, so it can safely store or modify it.
```
You're right, I guess it's not realistic to modify the plan, they just act on the config. I've removed the copy.

benhoyt commented on internal/overlord/servstate/handlers.go:
```diff
@@ -68,13 +68,16 @@ const (
+	stateBackoff          serviceState = "backoff"
+	stateExited           serviceState = "exited"
+	stateCheckTerminating serviceState = "terminating (after check failure)"
+	stateCheckKilling     serviceState = "killing (after check failure)"
```
Yeah, I was definitely on the fence about these too, and wondered if there's a better way. I've removed them now, and replaced the state they track with a `restarting bool` field. This makes the state machine simpler, though it does mean we have an additional state variable which is only valid in a couple of states. So I think strictly speaking those two additional states made sense as actual states, as they had different transitions/flow in the state machine, but pragmatically speaking, this may be simpler.

Note that the state machine is an implementation detail, not a list of user states. So these are to track internal states, not something that's exposed to the user.

Regarding `exiting`, that one's less clear -- it tracks the difference between manually stopped and the service just exiting itself. I tried removing it locally, but then you need to track another piece of state for "was manually stopped", because in `checkFailed` you don't want a check failure to backoff and restart a service if it's been manually stopped (only if it exited). Manual stop means stay stopped.

benhoyt replied:
Oh, I've also added a Graphviz diagram of the state machine. Not very nicely laid out, but here it is:

![state diagram](https://raw.githubusercontent.com/canonical/pebble/40e45233591c49618f20d85d3d91d6c3aa460e6f/internal/overlord/servstate/state-diagram.svg)

benhoyt commented on internal/overlord/servstate/handlers.go:
```diff
@@ -254,6 +257,10 @@ func (m *ServiceManager) serviceForStop(task *state.Task, name string) *serviceD
 		return nil
+	case stateExited:
+		taskLogf(task, "Service %q had already exited.", name)
+		service.transition(stateStopped)
```
Yes, in a few places we unify them, but in the one place we don't, it's a pain if it's not separate (we'd need an additional piece of state to track).

benhoyt commented on internal/overlord/servstate/handlers.go:
```diff
@@ -629,14 +644,18 @@ func (s *serviceData) terminateTimeElapsed() error {
+		if s.state == stateTerminating {
+			s.transition(stateKilling)
+		} else {
+			s.transition(stateCheckKilling)
```
Done.

benhoyt commented on internal/overlord/servstate/handlers.go:
```diff
@@ -658,6 +677,10 @@ func (s *serviceData) killTimeElapsed() error {
 
+	case stateCheckKilling:
+		logger.Noticef("Service %q still running after SIGTERM and SIGKILL", s.config.Name)
+		s.transition(stateStopped)
```
The `s.stopped` channel is only used (waited on) in `doStop` when the user is manually stopping the service, so you don't need/want to push to it in other cases.

benhoyt commented on internal/overlord/servstate/handlers.go:
```diff
@@ -442,6 +443,10 @@ func (s *serviceData) exited(waitErr error) error {
 
+	case stateCheckTerminating, stateCheckKilling:
+		logger.Noticef("Service %q exited after check failure, restarting", s.config.Name)
+		s.doBackoff(plan.ActionRestart, "on-check-failure")
```
As above - I removed these two new states.

benhoyt commented:
Ready for re-review. I've addressed most of your comments. With the exception of removing the `exited` state -- see comments above. Happy to chat about that -- though it'd be good if we can push ahead as this is an implementation detail that could go either way (doesn't change anything user-facing).

Regarding "shutdown" instead of "halt" -- Harry and I agree that makes sense in light of the other work there. Updated.

niemeyer reviewed (approved):
Thanks for all the changes, Ben!
